drv/transceivers-server/src/udp.rs

```diff
@@ -217,397 +215,588 @@ impl ServerImpl {
+        let response = Message::new(body);
+        let header = Header::new(msg_id, response.kind());
+
+        let hdr_len = hubpack::serialize(tx_data_buf, &header).unwrap();
+        let msg_len =
+            hubpack::serialize(&mut tx_data_buf[hdr_len..], &response).unwrap();
+
+        ringbuf_entry!(Trace::ResponseSize(ResponseSize {
+            header_length: hdr_len as u8,
+            message_length: msg_len as u8,
+            data_length: data_len as u16
+        }));
 
         // At this point, any supplementary data was written to
         // `tx_data_buf[Message::MAX_SIZE..]`, so it's not necessarily tightly
         // packed against the end of the `Message`.  Let's shift it backwards
         // based on the size of the leading `Message`:
         tx_data_buf.copy_within(
-            Message::MAX_SIZE..(Message::MAX_SIZE + data_len),
-            msg_len,
+            reserved_framing..(reserved_framing + data_len),
+            hdr_len + msg_len,
         );
-        Some((msg_len + data_len) as u32)
+        Some((hdr_len + msg_len + data_len) as u32)
     }
 
     fn handle_host_request(
         &mut self,
         h: HostRequest,
-        modules: ModuleId,
         data: &[u8],
         out: &mut [u8],
-    ) -> Result<(SpResponse, usize), Error> {
+    ) -> (MessageBody, usize) {
         match h {
-            HostRequest::Status => {
+            HostRequest::Status(modules) => {
                 ringbuf_entry!(Trace::Status(modules));
-                let count = self.get_status(modules, out)?;
-                Ok((SpResponse::Status, count))
+                let mask = LogicalPortMask::from(modules);
+                let (num_status_bytes, result) = self.get_status(mask, out);
+                ringbuf_entry!(Trace::OperationResult(result));
+                let success = ModuleId::from(result.success());
+                let (final_payload_len, failed_modules) = self.handle_errors(
+                    modules,
+                    result,
+                    None,
+                    num_status_bytes,
+                    out,
+                );
+
+                (
+                    MessageBody::SpResponse(SpResponse::Status {
+                        modules: success,
+                        failed_modules,
+                    }),
+                    final_payload_len,
+                )
             }
-            HostRequest::Read(mem) => {
-                ringbuf_entry!(Trace::Read(modules, mem));
-                let out_size = mem.len() as u32 * modules.ports.0.count_ones();
-                if out_size as usize > transceiver_messages::MAX_PAYLOAD_SIZE {
-                    return Err(Error::RequestTooLarge);
+            HostRequest::Read { modules, read } => {
+                ringbuf_entry!(Trace::Read(modules, read));
+                let mask = LogicalPortMask::from(modules);
+                let request_size = read.len() as u32 * mask.0.count_ones();
```

Comment:
What happens if I send a `HostRequestRead { modules: 0xFFFFFFFF_FFFFFFFF, read: MemoryRead { .. len: 32} }`?

It will success at this check, because `read.len() * mask.0.count_ones()` will be exactly 1024; however, the call to `handle_errors` below will _also_ try serializing a bunch of `HwError::InvalidModuleIndex` into the trailing buffer, resulting in a panic (I think!).

Reply:
Correct! My current understanding is that the host will not actually attempt to address the upper 32 bits right now, but obviously we'd like to gracefully handle should that occur. I could count how many bits are set in the upper 32, multiply that by `HwError::MAX_SIZE`, and then add that to `request_size` since we know for a fact we will be serializing errors there.

@bnaecker just a FYI in case you have anything to add here.

Reply:
The host won't try to set any of the upper 32 bits, but there's not much preventing that from happening in the future. Your idea of adding that to the request size and then possibly failing it sounds good.